snow/engine/snowman/transitive.go

```diff
@@ -492,52 +492,78 @@ func (t *Transitive) Context() *snow.ConsensusContext {
 
 func (t *Transitive) Start(ctx context.Context, startReqID uint32) error {
 	t.requestID = startReqID
-	lastAcceptedID, err := t.VM.LastAccepted(ctx)
+	preferredID, err := t.VM.GetPreference(ctx)
 	if err != nil {
 		return err
 	}
-
-	lastAccepted, err := t.GetBlock(ctx, lastAcceptedID)
+	preferredBlock, err := t.VM.GetBlock(ctx, preferredID)
 	if err != nil {
-		t.Ctx.Log.Error("failed to get last accepted block",
-			zap.Error(err),
-		)
 		return err
 	}
+	preferredChain := make([]snowman.Block, 0, 1)
+	for preferredBlock.Status() != choices.Accepted {
+		preferredChain = append(preferredChain, preferredBlock)
+
+		preferredBlock, err = t.VM.GetBlock(ctx, preferredBlock.Parent())
+		if err != nil {
+			return err
+		}
+	}
+	lastAccepted := preferredBlock
+	lastAcceptedID := lastAccepted.ID()
+	preferredChain = append(preferredChain, lastAccepted)
 
 	// initialize consensus to the last accepted blockID
 	if err := t.Consensus.Initialize(t.Ctx, t.Params, lastAcceptedID, lastAccepted.Height(), lastAccepted.Timestamp()); err != nil {
 		return err
 	}
 
-	// to maintain the invariant that oracle blocks are issued in the correct
-	// preferences, we need to handle the case that we are bootstrapping into an oracle block
-	if oracleBlk, ok := lastAccepted.(snowman.OracleBlock); ok {
-		options, err := oracleBlk.Options(ctx)
-		switch {
-		case err == snowman.ErrNotOracle:
-			// if there aren't blocks we need to deliver on startup, we need to set
-			// the preference to the last accepted block
-			if err := t.VM.SetPreference(ctx, lastAcceptedID); err != nil {
+	issuedMetric := t.metrics.issued.WithLabelValues(builtSource)
+	for len(preferredChain) > 0 {
```

Comment:
Instead of appending the last accepted block into the preference chain, I think it would be easier if we kept them separate to simplify the below logic:

1. This way we can just blindly throw all preferred blocks in-order into consensus without having to do the weird double trim on options blocks to guarantee we don't double-issue options into consensus
2. This way we can also handle the bootstrapping into oracle block case separately after handling the preferred blocks and issue the options as needed

Reply:
You're right. I implemented it this way to ensure there was at least one block, so that we could call `SetPreference` with the tip even when there were no preferred blocks that hadn't been accepted yet.

If we switch to relying on the VM to handle that as you mentioned here https://github.com/ava-labs/avalanchego/pull/2639/files/1ab3c85f2e044eba1d735e1cc0fb1e5847c1a90a#r1508507408, then that's not necessary and we can clean this up.

That being said, I think we still need to handle the options the same way? Definitely possible that I'm misunderstanding though cc @StephenButtolph

Reply:
Discussed offline, summarizing here:

For an oracle block in the preferred chain, we only need to insert the preferred child during restart. This means that we can completely skip this complexity. When we hit an oracle block, we will insert it and then insert the next block in the preferred chain, which is guaranteed to be the block we should prefer after restart. There is no need to guarantee that BOTH of the child blocks will be delivered on restart. The engine should handle that case gracefully if it ends up hearing about the other child block and/or switching its preference to the alternative child as needed.

Since the engine will always insert the children of an oracle block in the VM's pre-defined order, it also guarantees that it never calls `SetPreference` directly on the oracle block. Instead, it should always call `SetPreference` on one of the two children.

The only case we need to handle an oracle block's children here is for bootstrapping into the last accepted block as it was handled before.